axlearn/open_api/metrics/tool_use_execution.py

```diff
@@ -232,6 +249,128 @@ def _compare_tool_calls(
     return True
 
 
+@dataclasses.dataclass
+class DetailedMatchResult:
+    """Represents the tool matches for different metrics."""
+
+    func_name_match: bool = False
+    strict_arg_match: bool = False
+    lenient_arg_match: bool = False
+    lenient_bow_arg_match: bool = False
+
+
+def _compare_tool_call_detailed(
+    *,
+    pred_tool_calls: Sequence[dict[str, Any]],
+    target_tool_calls: Sequence[dict[str, Any]],
+) -> list[DetailedMatchResult]:
+    """Performs a detailed comparison of the predicted tool calls with the target tool calls and
+    returns different metrics.
+
+    The number of predicted and target tool calls can be different. The comparsion works by
+    iterating over all predicted calls and finding the corresponding target call. Correspondence
+    is defined here by the following rules:
+    - Strict metric: The function name matches and all arguments are identical.
+    - Lenient metric: The function name matches and all arguments match by the lenient argument
+      comparision.
+    - Lenient BOW metric: The function name matches and all arguments match by the lenient BOW
+      comparison.
+
+    Note, that this function return the matching results for every predicted tool call. The final
+    metric is taking the overall number of target calls into account too. That happens in the
+    calling function.
+
+    Args:
+        pred_tool_calls: Predicted tool calls.
+        target_tool_calls: Target tool calls.
+
+    Returns:
+        DetailedMatchResult, containing the Booleans denoting matches for different metrics.
```

Comment:
```suggestion
        A list of DetailedMatchResults with the same length as `pred_tool_calls`. Each result indicates whether the corresponding predicted tool call matches any target tool call.
```

Reply:
Correct. Good suggestion. That uncovered a small bug. Thanks.